### PR TITLE
Convert to TypeScript and Webpack #48

### PR DESCRIPTION
**Pull request by sankerust**
closes [placeholder]

Conversion to Typescript and Webpack.

Build ok:
image
image

Plugin itself:
image

image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,34 +19,38 @@
...
+    "style-loader": "^0.23.0",
+    "text-loader": "0.0.1",
+    "ts-loader": "^5.2.1",
+    "typescript": "^3.0.3",
+    "webpack": "^4.20.2",
+    "webpack-cli": "^3.1.1"
   },
   "dependencies": {
-    "lodash": "~4.0.0"
+    "css-loader": "^1.0.0",
```
Comment (rozetko) [inline package.json]: let's move it to dev-deps

File: src/datasource/config.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (rozetko) [inline src/datasource/config.ts]: let's remove empty first and leading lines

File: src/panel/device-list/module.ts [hunk 3]
```diff
@@ -12,12 +12,13 @@ var panelDefaults = {
 };
 
 class DeviceListCtrl extends PanelCtrl {
+  devices: any[];
+  pageReady: boolean;
+  static templateUrl: any;
```
Comment (jonyrock) [inline src/panel/device-list/module.ts]: lets put in first as in other files.

File: src/components/add_device.ts [hunk 4]
```diff
@@ -15,11 +15,12 @@ var defaults = {
 };
 
 export class AddDeviceCtrl {
+  device: any;
+  sending_ips: any[];
+  static templateUrl: string;
```
Comment (jonyrock) [inline src/components/add_device.ts]: lets put in first as in other files

File: src/components/device_details.ts [hunk 5]
```diff
@@ -1,10 +1,12 @@
 export class DeviceDetailsCtrl {
+  device: any;
+  deviceDTO: any;
+  pageReady: boolean;
+  other_ips: any;
+  static templateUrl: string;
```
Comment (jonyrock) [inline src/components/device_details.ts]: lets put in first as in other files

File: src/components/device_list.ts [hunk 6]
```diff
@@ -1,8 +1,10 @@
 class DeviceListCtrl {
+  devices: any[];
+  pageReady: boolean;
+  static templateUrl: string;
```
Comment (jonyrock) [inline src/components/device_list.ts]: lets put in first as in other files

File: src/panel/call-to-action/module.ts [hunk 7]
```diff
@@ -12,11 +12,13 @@ var panelDefaults = {
 };
 
 class CallToActiontCtrl extends PanelCtrl {
+  deviceStatus: string;
+  AllDone: boolean;
+  static templateUrl: string;
```
Comment (jonyrock) [inline src/panel/call-to-action/module.ts]: lets put in first as in other files

File: build/webpack.base.conf.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+function resolve(dir) {
+  return path.join(__dirname, '..', dir)
+}
+
+module.exports = {
+  target: 'node',
+  context: resolve('src'),
+  entry: {
+    './module': './module.ts',
```
[the remark this answers is not in the document]
Comment (sankerust) [reply build/webpack.base.conf.js]: Yes, there are different entries for each panel, datasource and for a few accessory files.

File: build/webpack.base.conf.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    'panel/call-to-action/module': './panel/call-to-action/module.ts',
+    'panel/device-list/module': './panel/device-list/module.ts',
+  },
+  output: {
+    filename: "[name].js",
+    path: resolve('dist'),
+    libraryTarget: "amd"
+  },
+  externals: [
+    // remove the line below if you don't want to use buildin versions
```
[the remark this answers is not in the document]
Comment (sankerust) [reply build/webpack.base.conf.js]: Yes, it is.

File: build/webpack.base.conf.js [hunk 10]
```diff
@@ -0,0 +1,98 @@
+const path = require('path');
+const webpack = require('webpack');
+const CopyWebpackPlugin = require('copy-webpack-plugin');
+const CleanWebpackPlugin = require('clean-webpack-plugin');
+const ExtractTextPlugin = require('extract-text-webpack-plugin');
+
+const ngAnnotatePlugin = require('ng-annotate-webpack-plugin');
```
[the remark this answers is not in the document]
Comment (sankerust) [reply build/webpack.base.conf.js]: It is needed when running npm run dev

File: build/webpack.base.conf.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (jonyrock) [reply build/webpack.base.conf.js]: @rozetko think about ... :) Cuz here you can define your logic your scss files

File: src/datasource/kentikProxy.ts [hunk 11]
```diff
@@ -30,13 +30,13 @@ function getMaxRefreshInterval(query) {
 
 class KentikProxy {
   kentikAPI: any;
-  cache: {};
+  cache: any;
   cacheUpdateInterval: number;
   requestCachingIntervals: { '1d': number; };
-  getDevices: any;
+  getDevices: Function;
```
Comment (jonyrock) [inline src/datasource/kentikProxy.ts]: A bit more specific https://www.typescriptlang.org/docs/handbook/functions.html
See "Writing function types"

File: src/datasource/datasource.ts [hunk 12]
```diff
@@ -1,15 +1,16 @@
 import {metricList, unitList, filterFieldList} from './metric_def';
 import _ from 'lodash';
-import TableModel from 'app/core/table_model';
+import TableModel from 'grafana/app/core/table_model';
 import './kentikProxy';
```
[the remark this answers is not in the document]
Comment (sankerust) [reply src/datasource/datasource.ts]: done

File: src/components/add_device.ts (hunk 4 above)
Comment (sankerust) [reply src/components/add_device.ts]: done

File: src/panel/call-to-action/module.ts (hunk 7 above)
Comment (sankerust) [reply src/panel/call-to-action/module.ts]: done

**Pull request closed without merging**